ambry-account/src/main/java/com/github/ambry/account/RouterStore.java

```diff
@@ -0,0 +1,331 @@
+          }
+        }
+      }
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  private Map<String, String> readAccountMetadataFromBlobID(String blobID) {
+    long startTimeMs = System.currentTimeMillis();
+    Future<GetBlobResult> resultF = accountService.router.get().getBlob(blobID, new GetBlobOptionsBuilder().build());
+    try {
+      GetBlobResult result = resultF.get();
+      accountService.accountServiceMetrics.accountFetchFromAmbryTimeInMs.update(
+          System.currentTimeMillis() - startTimeMs);
+
+      int blobSize = (int) result.getBlobInfo().getBlobProperties().getBlobSize();
+      InputStream input = new ReadableStreamChannelInputStream(result.getBlobDataChannel());
+      byte[] bytes = Utils.readBytesFromStream(input, blobSize);
+
+      JSONObject object = new JSONObject(new String(bytes, Charsets.UTF_8));
+      Map<String, String> map = new HashMap<>();
+      object.keySet().stream().forEach(key -> map.put(key, object.getString(key)));
+      return map;
+    } catch (Exception e) {
+      logger.debug("Failed to read account metadata from blob id={}", blobID, e);
+      accountService.accountServiceMetrics.accountFetchFromAmbryServerErrorCount.inc();
+    }
+    return null;
+  }
+
+  @Override
+  public boolean updateAccounts(Collection<Account> accountsToUpdate) {
+    ZKUpdater zkUpdater = new ZKUpdater(accountsToUpdate);
+    boolean hasSucceeded = helixStore.update(ACCOUNT_METADATA_BLOB_IDS_PATH, zkUpdater, AccessOption.PERSISTENT);
+    zkUpdater.cleanup(hasSucceeded);
+    return hasSucceeded;
+  }
+
+  static String writeAccountMapToRouter(Map<String, String> accountMap, Router router) throws Exception {
+    // Construct the json object and save it to ambry server.
+    JSONObject object = new JSONObject();
+    for (Map.Entry<String, String> entry : accountMap.entrySet()) {
+      object.put(entry.getKey(), entry.getValue());
+    }
+
+    ByteBufferReadableStreamChannel channel =
+        new ByteBufferReadableStreamChannel(ByteBuffer.wrap(object.toString().getBytes(Charsets.UTF_8)));
+    BlobProperties properties = new BlobProperties(channel.getSize(), SERVICE_ID, ACCOUNT_ID, CONTAINER_ID, false);
+    return router.putBlob(properties, null, channel, PutBlobOptions.DEFAULT).get();
+  }
+
+  private class ZKUpdater implements DataUpdater<ZNRecord> {
+    private final Collection<Account> accounts;
+    private final Pair<String, Path> backupPrefixAndPath;
+    private Map<String, String> potentialNewState;
+    private String newBlobID = null;
+
+    ZKUpdater(Collection<Account> accounts) {
+      this.accounts = accounts;
+      Pair<String, Path> backupPrefixAndPath = null;
+      if (accountService.backupDirPath != null) {
+        try {
+          backupPrefixAndPath = accountService.reserveBackupFile();
+        } catch (IOException e) {
+          logger.error("Error reserving backup file", e);
```

Comment:
What happens later if the backup file wasn't created?  We skip the backups?

Reply:
With the current logic, yes.
The current backup file logic is subpar at best. It doesn't guarantee the atomicity of writes and doesn't clean up old files. But I am using the same set of tests for the new implementation so I don't want to change the backup logic. Will make it better at later PR.

Reply:
Okay.  I don't like the code but I can see you didn't change it only moved it around.